test/resources/PureScmTest.Constructor.purs

```diff
@@ -0,0 +1,10 @@
+module PureScmTest.Constructor where
+
+data Foo = Bar Int Int
+         | Baz
+
+match (Bar 1 2)   = 0
+match (Bar 1 _)   = 1
+match (Bar _ 2)   = 2
+match (Bar i1 i2) = i1
+match Baz = 0
```

Comment:
Is this test exercising all the code paths?

Reply:
In theory yes: I ran `stack test --coverage` and `stack hpc report .` and it says that the only never executed paths are the `error` placeholders.

In practice: we should test nested constructor even though I tested your previous test case and it seems to work.

Reply:
Let's add nested constructors to this test, and let's try to hit all the `error` placeholders

Reply:
I will add another test case (NestedConstructor.purs) with the following code

```
data Foo = Foo Int

data Bar = Bar Foo

match :: Bar -> Int
match (Bar (Foo 0)) = 1
match _ = 23
```

Let me know if I have to change anything.

Reply:
Perfect!

Reply:
The JS compiler translates that code into

```
var Foo = (function () {
    function Foo(value0) {
        this.value0 = value0;
    };
    Foo.create = function (value0) {
        return new Foo(value0);
    };
    return Foo;
})();

var Bar = (function () {
    function Bar(value0) {
        this.value0 = value0;
    };
    Bar.create = function (value0) {
        return new Bar(value0);
    };
    return Bar;
})();

var match = function (v) {
    if (v.value0.value0 === 0) {
        return 1;
    };
    return 23;
};
```

while the Scheme compiler does this

```
(define Foo
  (lambda (value0)
    (cons (quote Foo) (vector value0))))

(define Bar
  (lambda (value0)
    (cons (quote Bar) (vector value0))))

(define match
  (lambda (v)
    (cond ((and (eq? (car v) (quote Bar))
                (eq? (car (vector-ref (cdr v) 0)) (quote Foo))
                (= (vector-ref (cdr (vector-ref (cdr v) 0)) 0) 0)) 1)
          (#t 23))))
```

Are we doing way too many tests?

Reply:
Discussion about this moved to #27